### ISSUE TITLE
Add pod failed due to addLogicalPort error

### ISSUE THREAD

**Issue by amoy-xuhao** [ovn-org/ovn-kubernetes#2964]
ovnkube-master logs:
I0503 05:28:39.009294       1 ovn.go:460] [b7e15390-d6ce-4561-a40d-158b9cc788c6/openshift-multus/network-metrics-daemon-tgvcc] retry pod setup
I0503 05:28:39.009546       1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.3/23"],"mac_address":"0a:58:0a:80:00:03","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.3/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc
I0503 05:28:39.009605       1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.4/23"],"mac_address":"0a:58:0a:80:00:04","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.4/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc
2022-05-03T05:28:39.011Z|00019|ovn_dbctl|INFO|Running command run -- set nb_global . options:controller_event=true
I0503 05:28:39.019747       1 pods.go:282] [openshift-multus/network-metrics-daemon-tgvcc] addLogicalPort took 10.435665ms, OVN Execute time 1.919191ms, pod Annotation time: 8.205885ms
I0503 05:28:39.019765       1 ovn.go:463] [b7e15390-d6ce-4561-a40d-158b9cc788c6/openshift-multus/network-metrics-daemon-tgvcc] pod setup successful
I0503 05:28:39.025459       1 repair.go:110] Deleted 0 stale service LBs
I0503 05:28:39.029462       1 pods.go:282] [openshift-multus/network-metrics-daemon-tgvcc] addLogicalPort took 20.319581ms, OVN Execute time 940.075µs, pod Annotation time: 18.96604ms
E0503 05:28:39.029479       1 ovn.go:560] error while creating logical port openshift-multus_network-metrics-daemon-tgvcc error: Reconnecting...Transaction Failed due to an error: constraint violation details: Transaction causes multiple rows in "Logical_Switch_Port" table to have identical values (openshift-multus_network-metrics-daemon-tgvcc) for index on column "name".  First row, with UUID cf022291-08e3-4ace-9e6b-7c1179d97205, existed in the database before this transaction and was not modified by the transaction.  Second row, with UUID 2fb9e95b-d937-45f4-abe4-cfaaab0284f0, was inserted by this transaction. in committing transaction
I0503 05:28:39.029505       1 ovn.go:650] Bootstrapping existing pods and cleaning stale pods took 20.619049ms

**Comment by tssurya** [comment]
Hi @amoy-xuhao  thanks for opening the issue!
Could you please paste the OCP version you were using when you hit this error?

**Comment by amoy-xuhao** [comment]
@tssurya Thank you for your attention
The OCP version is 4.9.0

**Comment by amoy-xuhao** [comment]
In the ensurePod function, a lock should be required; because when the ovnkube-master controller starts, the RetryPods goroutine starts, and at the same time, WatchPods will also execute the ensurePod function

**Comment by tssurya** [comment]
> E0503 05:28:39.029479 1 ovn.go:560] error while creating logical port openshift-multus_network-metrics-daemon-tgvcc error: Reconnecting...Transaction Failed due to an error: constraint violation details: Transaction causes multiple rows in "Logical_Switch_Port" table to have identical values (openshift-multus_network-metrics-daemon-tgvcc) for index on column "name". First row, with UUID cf022291-08e3-4ace-9e6b-7c1179d97205, existed in the database before this transaction and was not modified by the transaction. Second row, with UUID 2fb9e95b-d937-45f4-abe4-cfaaab0284f0, was inserted by this transaction. in committing transaction

This is just a red-herring if a monitor drops silently we will force a reconnect: https://bugzilla.redhat.com/show_bug.cgi?id=1998614#c15

**Comment by amoy-xuhao** [comment]
> > E0503 05:28:39.029479 1 ovn.go:560] error while creating logical port openshift-multus_network-metrics-daemon-tgvcc error: Reconnecting...Transaction Failed due to an error: constraint violation details: Transaction causes multiple rows in "Logical_Switch_Port" table to have identical values (openshift-multus_network-metrics-daemon-tgvcc) for index on column "name". First row, with UUID cf022291-08e3-4ace-9e6b-7c1179d97205, existed in the database before this transaction and was not modified by the transaction. Second row, with UUID 2fb9e95b-d937-45f4-abe4-cfaaab0284f0, was inserted by this transaction. in committing transaction
> 
> This is just a red-herring if a monitor drops silently we will force a reconnect: https://bugzilla.redhat.com/show_bug.cgi?id=1998614#c15 It's kind of an expected fix/log

The problem is not that the monitor drops silently, but that the ovnkube-master controller has added two ips to the pod, causing the execution of the ovn command to fail

**Comment by amoy-xuhao** [comment]
@tssurya  

I0503 05:28:39.009546 1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.3/23"],"mac_address":"0a:58:0a:80:00:03","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.3/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc
I0503 05:28:39.009605 1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.4/23"],"mac_address":"0a:58:0a:80:00:04","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.4/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc

**Comment by tssurya** [comment]
> @tssurya
> 
> I0503 05:28:39.009546 1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.3/23"],"mac_address":"0a:58:0a:80:00:03","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.3/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc I0503 05:28:39.009605 1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.4/23"],"mac_address":"0a:58:0a:80:00:04","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.4/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc

yes but did you check the comment I linked on the bugzilla which points to another bugzilla as a consequence of that: https://bugzilla.redhat.com/show_bug.cgi?id=2002372#c1 which is what you see in this case?

Bottom line is, the code upstream has diverged by a LOT from OCP 4.9, are you seeing the same problem on a master branch?

**Comment by tssurya** [comment]
@amoy-xuhao : also note that you don't have to open and close PRs like you are doing currently to fix an issue, and instead if you have an update you can simply force push to the same branch and that will update a PR.

**Comment by amoy-xuhao** [comment]
> > @tssurya
> > I0503 05:28:39.009546 1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.3/23"],"mac_address":"0a:58:0a:80:00:03","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.3/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc I0503 05:28:39.009605 1 kube.go:67] Setting annotations map[k8s.ovn.org/pod-networks:{"default":{"ip_addresses":["10.128.0.4/23"],"mac_address":"0a:58:0a:80:00:04","gateway_ips":["10.128.0.1"],"ip_address":"10.128.0.4/23","gateway_ip":"10.128.0.1"}}] on pod openshift-multus/network-metrics-daemon-tgvcc
> 
> yes but did you check the comment I linked on the bugzilla which points to another bugzilla as a consequence of that: https://bugzilla.redhat.com/show_bug.cgi?id=2002372#c1 which is what you see in this case?
> 
> Bottom line is, the code upstream has diverged by a LOT from OCP 4.9, are you seeing the same problem on a master branch?

master branch has the problem too

**Comment by amoy-xuhao** [comment]
The logic of the master branch here has not changed

**Comment by amoy-xuhao** [comment]
The reason for the problem is that when the ovnkube-master leader starts, watchNodes will obtain all pods on the corresponding node and execute the ensurePod function; at the same time, watchPods will also operate the corresponding pod and execute the ensurePod function, there is a concurrency problem here

**Comment by tssurya** [comment]
> @tssurya The reason for the problem is that when the ovnkube-master leader starts, watchNodes will obtain all pods on the corresponding node and execute the ensurePod function; at the same time, watchPods will also operate the corresponding pod and execute the ensurePod function, there is a concurrency problem here

do you have a cluster live where this is happening on master? The multus pod logs you are showing are not on master, they are from 4.9.

**Comment by amoy-xuhao** [comment]
emmm...
I don‘t have the cluster with master branch；And the bug is hard to recurrent

**Comment by amoy-xuhao** [comment]
But the logic of concurrency handling ensurePod in master branch is the same as 4.9
